cnf-tests/testsuites/e2esuite/multinetworkpolicy/reachmatcher.go

```diff
@@ -0,0 +1,325 @@
+//  y/b   X       X       X       X       .       X       X       X       X
+//  y/c   X       X       X       X       X       .       X       X       X
+//  z/a   X       X       X       X       X       X       .       X       X
+//  z/b   X       X       X       X       X       X       X       .       X
+//  z/c   X       X       X       X       X       X       X       X       .
+func makeConnectivityMatrix(destinationPort string, protocol corev1.Protocol, pods ...*corev1.Pod) string {
+
+	type connectivityPair struct {
+		from  *corev1.Pod
+		to    *corev1.Pod
+		value bool
+	}
+
+	data := make(chan connectivityPair, 81)
+
+	connectivityMatrix := make(map[*corev1.Pod]map[*corev1.Pod]bool)
+
+	var conversionWG sync.WaitGroup
+	conversionWG.Add(1)
+	go func() {
+		defer conversionWG.Done()
+		for k := range data {
+			from, ok := connectivityMatrix[k.from]
+			if !ok {
+				from = make(map[*corev1.Pod]bool)
+				connectivityMatrix[k.from] = from
+			}
+			from[k.to] = k.value
+		}
+	}()
+
+	var producerWG sync.WaitGroup
+
+	for _, source := range pods {
+		for _, destination := range pods {
+			producerWG.Add(1)
+			d := destination
+			s := source
+			go func() {
+				defer producerWG.Done()
+
+				if s == nil || d == nil {
+					return
+				}
+				canReach, err := canSendTraffic(s, d, destinationPort, protocol)
+				if err != nil {
+					fmt.Println(err.Error())
+					return
+				}
+
+				data <- connectivityPair{from: s, to: d, value: canReach}
+			}()
+		}
+	}
+
+	producerWG.Wait()
+	close(data)
+	conversionWG.Wait()
+
+	return convertToString(destinationPort, protocol, connectivityMatrix, pods...)
+}
+
+func convertToString(destinationPort string, protocol corev1.Protocol, connectivityMatrix map[*corev1.Pod]map[*corev1.Pod]bool, pods ...*corev1.Pod) string {
+
+	ret := fmt.Sprintf("Reachability matrix of %d pods on %s:%s (X = true, . = false)\n", len(pods), protocol, destinationPort)
+
+	for _, destination := range pods {
+		ret += fmt.Sprintf("\t%s", shortName(destination))
+	}
+	ret += "\n"
+
+	for _, source := range pods {
+		ret += shortName(source)
+		from, ok := connectivityMatrix[source]
+		if !ok {
+			ret += "\n"
+			continue
+		}
+		for _, destination := range pods {
+			ret += "\t"
+			canReach, ok := from[destination]
+			if !ok {
+				ret += "?"
+				continue
+			}
+			if canReach {
+				ret += "X"
+				continue
+			}
+
+			ret += "."
+		}
+		ret += "\n"
+	}
+
+	return ret
+}
+
+func shortName(p *corev1.Pod) string {
+	ns := ""
+	switch p.Namespace {
+	case nsX:
+		ns = "x"
+	case nsY:
+		ns = "y"
+	case nsZ:
+		ns = "z"
+	}
+
+	podLabel, ok := p.ObjectMeta.Labels["pod"]
+	if !ok {
+		podLabel = "?"
+	}
+	return ns + "/" + podLabel
+}
+
+func getIPTables(pod *corev1.Pod) string {
+	output, err := pods.ExecCommand(client.Client, *pod, []string{"iptables", "-L", "-v", "-n"})
+	if err != nil {
+		return "<err: " + err.Error() + ">"
+	}
+
+	return output.String()
+}
+
+func getFirstSriovNicIP(pod *corev1.Pod) string {
```

Comment:
how this will work with dual stack?

Reply:
Good point. I can change the reachability logic to try to connect to all IP address assigned to an interface. Reachability would return true if and only if all IP address on the server can be reached by the client, false otherwise. 

Does it make sense?

Reply:
Currently, IPv6 is not supported by multus network policy iptables implementation. In any case, I added a skipped test and the logic for testing in a dual-stack environment.